server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+# Block Verification Design
+
+## Table of Contents
+
+1. [Purpose](#purpose)
+2. [Goals](#goals)
+3. [Terms](#terms)
+4. [Entities](#entities)
+5. [Design](#design)
+6. [Interfaces](#interfaces)
+7. [Enums](#enums)
+8. [Metrics](#metrics)
+9. [Exceptions](#exceptions)
+
+## Purpose
+
+The purpose of the Block Verification feature is to ensure that blocks received from consensus nodes are valid and have not been tampered with. This is achieved by re-calculating the block hash and verifying it against the signature provided by the consensus node.
+
+## Goals
+
+1) The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature.
+2) If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.
+
+
+## Terms
+
+- Consensus Node (CN): A node that produces and provides blocks.
+- Block Items: The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.
+- Block Hash: A cryptographic hash representing the block’s integrity.
+- Signature: The cryptographic signature of the block hash created by Network private key (aka: LedgerId)
+- Public Key: The public key of the LedgerId that signed the block.
+
+
+## Entities
+
+- **VerificationHandler** 
+  - Receives the stream of block items in the form of List<BlockItemsUnparsed> from the unparsed and unverified block items ring buffer. 
+  - When it detects a block_header, it creates a BlockHashingSession using the BlockHashingSessionFactory, providing it with the initial block items (and internally, the session will handle asynchronous hashing).
+  - Adds subsequent block items to the session, including the block_proof.
+  - Once the block_proof is received, calls completeHashing() on the BlockHashingSession, then releases the thread back to the UnverifiedRingBuffer.
+  - Does not block waiting for verification; the hash computation and verification continue asynchronously.
+- **BlockHashingSessionFactory**
+  - Creates new BlockHashingSession instances, provides them with a ExecutorService.
+- **BlockHashingSession**
+  - Holds all necessary block data.
+  - Accepts block items incrementally. (continues to compute them asynchronously)
+  - Once the block_proof is provided, finalizes the hash computation asynchronously.
+  - After computing the final hash, calls SignatureVerifier for verification.
+- **SignatureVerifier**
+  - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).
+  - Calls BlockStatusManager with the verification result.
+- **BlockStatusManager**
+  - Receives verification results from SignatureVerifier.
+  - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.
+
+## Design
+
+1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.
```

Comment:
Since we are getting a list of items, that is the same interface that we are using for the rest of the services. and other services. I believe that this is related to the definition of interfaces for services/modules of the system and I think we should take it out of this scope and discuss it independent of this Epic, since we already have an interface in place.

Reply:
Same, moving this to another [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400) so we can tackle the communication interface refactor on a separate Epic.

Reply:
I don't think this is the same topic as that discussion.
Here I think we have an incorrect interface definition, rather than a more broad "how do we have independent services" discussion.

Reply:
Have made the pertaining corrections on the design. 👍

Reply:
We're receiving a stream of collections of BlockItems from the CN. We're emitting a stream of collections of BlockItems to consumers. I'm confused by the proposal to loop through the inbound collection to broadcast single items on the ring buffer. This is just added looping computation when the goal is to get the data as quickly as possible to the components that require them.

Reply:
I believe we should not be requiring the _same_ level of batching at every step.
CN might send 100 at a time, then Verification sends 1000 at a time to hashing, and each subscriber makes a different request for batch size based on their own time/overhead tradeoffs.
Sending everything internally in the same batch as we receive removes all flexibility and control and puts it at "whatever the publisher sent" with no ability to change that.

We _may_ need to be careful how often we send data between threads (that's a tricky topic and we need to test the impact in our system, rather than blindly following what Consensus Nodes experience in their much more complex system), but I think we can be reasonable to batch things up to an appropriate level _where needed_ and by sending everything into the ring buffer individually we gain both greater simplicity (the ring buffer is always single items, regardless of external entities) and more flexibility to control any internal batch sizes.

We also get the advantage that we aren't looping through lists in _every_ service trying to detect headers and proof items; we just handle each item as it comes in and if it's a header or proof then we know a block just started or ended.

So we trade one loop on intake (where it's simple and easy, and there really isn't anything else for that component to do) for multiple loops in other services, which seems a good trade in my view.

Reply:
great discussion @mattp-swirldslabs and @jsync-swirlds, however we already created a place for that.

I agree that it implicates design changes to the VerificationHandler, but it also would do so for the rest of the current architecture, so for the time being we decided on take out of the current scope, and move it to a discussion #400 where we can come to an agreement and then create an EPIC for doing the necessary refactors.